modules/standard/List.chpl

```diff
@@ -144,17 +180,121 @@ record list {
       delete current;
       current = next;
     }
+    first = nil;
+    last = nil;
+    length = 0;
   }
 
   pragma "no doc"
   proc writeThis(f) {
-    var first: bool = true;
+    var binary = f.binary();
+    var arrayStyle = f.styleElement(QIO_STYLE_ELEMENT_ARRAY);
+    var isspace = arrayStyle == QIO_ARRAY_FORMAT_SPACE && !binary;
+    var isjson = arrayStyle == QIO_ARRAY_FORMAT_JSON && !binary;
+    var ischpl = arrayStyle == QIO_ARRAY_FORMAT_CHPL && !binary;
+
+    if binary {
+      // Write the number of elements.
+      f <~> length;
+    }
+    if isjson || ischpl {
+      f <~> new ioLiteral("[");
+    }
+
+    var first = true;
     for e in this {
-      if !first then
-        f.write(" ");
-      else
-        first = false;
-      f.write(e);
+      if first then first = false;
+      else {
+        if isspace then f <~> new ioLiteral(" ");
+        else if isjson || ischpl then f <~> new ioLiteral(",");
+      }
+
+      f <~> e;
+    }
+
+    if isjson || ischpl {
+      f <~> new ioLiteral("]");
+    }
+
+  }
+
+  pragma "no doc"
+  proc readThis(f) {
+    // Special handling for reading in order to handle
+    // reading an arbitrary length.
+    var binary = f.binary();
+    var arrayStyle = f.styleElement(QIO_STYLE_ELEMENT_ARRAY);
+    var isspace = arrayStyle == QIO_ARRAY_FORMAT_SPACE && !binary;
+    var isjson = arrayStyle == QIO_ARRAY_FORMAT_JSON && !binary;
+    var ischpl = arrayStyle == QIO_ARRAY_FORMAT_CHPL && !binary;
+
+    // How many elements should we read (for binary mode)?
+    var num = 0;
+
+    if binary {
+      // Read the number of elements.
+      f <~> num;
+    }
+    if isjson || ischpl {
+      f <~> new ioLiteral("[");
+    }
+
+    if ! f.error() {
+      // Clear out existing elements in the list
+      destroy();
+    }
+
+    var first = true;
+    var i = 0;
+    var read_end = false;
+
+    while ! f.error() {
+      if binary {
+        // Read only num elements.
+        if i >= num then break;
+      } else {
+        if first {
+          first = false;
+          // but check for a ]
+          if isjson || ischpl {
+            f <~> new ioLiteral("]");
+          } else if isspace {
+            f <~> new ioNewline(skipWhitespaceOnly=true);
+          }
+          if f.error() == EFORMAT {
+            f.clearError();
+          } else {
+            read_end = true;
+            break;
+          }
+        } else {
+          // read a comma or a space.
+          if isspace then f <~> new ioLiteral(" ");
+          else if isjson || ischpl then f <~> new ioLiteral(",");
```

Comment:
In the reader, we don't want to require the space. ioLiteral defaults to consuming extra space.